src/opencl/mod.rs

```diff
@@ -201,121 +173,174 @@ fn get_device_by_index(index: usize) -> Option<&'static Device> {
 }
 
 pub struct Program {
-    device: Device,
-    program: ocl::Program,
-    queue: ocl::Queue,
+    device_name: String,
+    queue: CommandQueue,
+    context: Context,
+    kernels_by_name: HashMap<String, opencl3::kernel::Kernel>,
 }
 
 impl Program {
-    pub fn device(&self) -> Device {
-        self.device.clone()
+    pub fn device_name(&self) -> &str {
+        &self.device_name
     }
-    pub fn from_opencl(device: Device, src: &str) -> GPUResult<Program> {
-        let cached = utils::cache_path(&device, src)?;
+
+    pub fn from_opencl(device: &Device, src: &str) -> GPUResult<Program> {
+        let cached = utils::cache_path(device, src)?;
         if std::path::Path::exists(&cached) {
             let bin = std::fs::read(cached)?;
             Program::from_binary(device, bin)
         } else {
-            let context = ocl::Context::builder()
-                .platform(device.platform)
-                .devices(device.device)
-                .build()?;
-            let program = ocl::Program::builder()
-                .src(src)
-                .devices(ocl::builders::DeviceSpecifier::Single(device.device))
-                .build(&context)?;
-            let queue = ocl::Queue::new(&context, device.device, None)?;
+            let context = Context::from_device(&device.device)?;
+            let mut program = opencl3::program::Program::create_from_source(&context, src)?;
+            if let Err(build_error) = program.build(context.devices(), "") {
+                let log = program.get_build_log(context.devices()[0])?;
+                return Err(GPUError::Opencl3(build_error, Some(log)));
+            }
+            let queue = CommandQueue::create(&context, context.default_device(), 0)?;
+            let kernels = opencl3::kernel::create_program_kernels(&program)?;
+            let kernels_by_name = kernels
+                .into_iter()
+                .map(|kernel| {
+                    let name = kernel.function_name()?;
+                    Ok((name, kernel))
+                })
+                .collect::<Result<_, ClError>>()?;
             let prog = Program {
-                program,
+                device_name: device.name(),
                 queue,
-                device,
+                context,
+                kernels_by_name,
             };
-            std::fs::write(cached, prog.to_binary()?)?;
+            let binaries = program
+                .get_binaries()
+                .map_err(|_| GPUError::ProgramInfoNotAvailable(CL_PROGRAM_BINARIES))?;
+            std::fs::write(cached, binaries[0].clone())?;
```

Comment:
what if there are multiple or no binary?

Reply:
lets open an issue, to at least investigate this

Reply:
https://github.com/filecoin-project/rust-gpu-tools/issues/43